pilgen/src/batcher.rs

```diff
@@ -0,0 +1,206 @@
+use std::{
+    collections::{BTreeMap, HashSet},
+    marker::PhantomData,
+};
+
+use itertools::Itertools;
+use number::FieldElement;
+use parser::{
+    asm_ast::{
+        batched::{ASMStatementBatch, BatchedASMFile, Incompatible},
+        ASMFile, ASMStatement, InstructionBodyElement, RegisterFlag,
+    },
+    ast::Expression,
+    expr_any,
+};
+
+#[derive(Default)]
+pub struct ASMBatcher<T> {
+    pc_name: Option<String>,
+    instruction_uses_pc: BTreeMap<String, bool>,
+    marker: PhantomData<T>,
+}
+
+impl<T: FieldElement> ASMBatcher<T> {
+    fn compatible(
+        &self,
+        left: &ASMStatement<T>,
+        right: &ASMStatement<T>,
+    ) -> Result<(), Incompatible> {
+        match (left, right) {
+            // we can always combine a label with the next statement
+            (ASMStatement::Label(..), _) => Ok(()),
+            // we cannot combine anything with a label
+            (_, ASMStatement::Label(..)) => Err(Incompatible::Label),
+            // we can combine some assignments
+            (
+                ASMStatement::Assignment(_, write_regs_0, assign_reg_0, value_0),
+                ASMStatement::Assignment(_, write_regs_1, assign_reg_1, value_1),
+            ) => {
+                // we're writing to different registers
+                if write_regs_0
+                    .iter()
+                    .collect::<HashSet<_>>()
+                    .is_disjoint(&write_regs_1.iter().collect::<HashSet<_>>())
+                {
+                } else {
+                    return Err(Incompatible::BusyWriteRegister);
+                }
+                // we're using different assignment registers
+                if assign_reg_0 != assign_reg_1 {
+                } else {
+                    return Err(Incompatible::BusyAssignmentRegister);
+                }
+                // the registers read by the first assignment are not written to by the second
+                // note: the opposite is fine: `A = 2; B = A`
+                // also, the second assignment does not write to registers it reads
+                if !expr_any(value_0, |e: &Expression<T>| match e {
+                    Expression::PolynomialReference(pol) => write_regs_1.contains(&pol.name),
+                    _ => false,
+                }) && !expr_any(value_1, |e: &Expression<T>| match e {
+                    Expression::PolynomialReference(pol) => write_regs_1.contains(&pol.name),
+                    _ => false,
+                }) {
+                } else {
+                    return Err(Incompatible::WriteAfterRead);
+                }
+                // if we are making function calls, we cannot use the pc // TODO allow a single static jump per batch
```

Comment:
I don't see the connection between function calls and the pc

Reply:
Changed to checking the whole expression